In [4]:
import pandas as pd
import csv
import os
import string
from datetime import datetime
import pickle
import re
from tqdm import tqdm
import gzip


In [ ]:
# 나중에 참고해볼거 :https://towardsdatascience.com/part-1-data-cleaning-does-bert-need-clean-data-6a50c9c6e9fd

In [2]:
pwd

'C:\\Users\\Samsung\\jupyter\\Project_Medical_Legal'

In [24]:
path = os.chdir(r"C:\Users\Samsung\jupyter\Project_Medical_Legal")

In [27]:
with open('./health_news_inputdata.pkl', 'rb') as f: 
    all_article = pickle.load(f)

AttributeError: Can't get attribute '_unpickle_block' on <module 'pandas._libs.internals' from 'C:\\Users\\Samsung\\AppData\\Roaming\\Python\\Python37\\site-packages\\pandas\\_libs\\internals.cp37-win_amd64.pyd'>

In [ ]:
with gzip.open('testPickleFile.pickle','rb') as f:
    data = pickle.load(f)

In [13]:
with open('health_news_inputdata.plk', 'rb') as f:
    data = pickle.load(f)

UnpicklingError: invalid load key, '\x1f'.

In [ ]:
with open('list.txt', 'rb') as f:
    data = pickle.load(f) 

In [970]:
title = []
body = []
days = []
error = []

for path in tqdm(paths):
        with open(path, newline='') as f : 
            file = csv.reader(f, delimiter='\t')
            day = datetime.strptime(path.split('.')[0], '%Y%m%d')

            for idx, article in enumerate(file):
                if len(article) < 2:
                    article = article[0].split('\t')
                try:
                    body.append(article[1])
                    title.append(article[0])
                    
                    days.extend([day])

                except:
                    error.append(article)
                    pass
                
                


100%|██████████████████████████████████████████████████████████████████████████████| 6726/6726 [00:07<00:00, 937.92it/s]


In [979]:
len(title)

530480

In [977]:
body[:10]

['[뉴스데스크]● 앵커： 요즘 우리 평균수명은 많이 늘었지만 몸이 불편해 고생하시는 노인들 주변에 많이 봅니다. 그냥 오래 사는 게 아니라 건강하게 오래 사는 이른바 건강수명이 중요하죠. 우리의 현 실태를 임영서 기자가 집중 취재했습니다. 한국인의 평균수명은 현재 78. 5세, OECD 국가의 평균치인 79세를 거의 따라 잡았습니다. 그러나, 사고나 질병으로 거동이 불편한 기간을 뺀 "건강히 사는 기간", 즉 "건강수명"은 그리 높지 않습니다. 최근 조사에 따르면 건강수명은 68. 6세, 즉 말년에 10년 가량은 병상에 있다가 삶을 마치는 셈입니다. ● 강은정 박사(보건사회연구원) : "일반대중을 위해 장애를 예방한다든지 장애상태에서 일상생활에 활동성을 높인다든지 그런 관심을 갖는 건 최근이었다. " 우리의 건강수명은 30개 OECD 국가가운데 24위로, 우리보다 못한 나라는 체코 헝가리 등 동구권 국가와 터키 멕시코뿐입니다. 덴마크 미국 등은 우리보다 평균수명은 짧지만 건강수명은 오히려 앞섭니다. 골다공증을 앓다가 허리 골절로 입원한 할머니입니다. 골절이 심해 치료를 마쳐도 정상적인 상태로 돌아가기는 어려운 실정입니다.● 이말순(68세, 골다공증 환자) : “여섯 대가 나갔다든가... 그때는 무척 아팠어요.” 이처럼 노인들의 골다공증은 골절로 이어질 우려가 크기 때문에 초기 치료와 관리가 절실합니다. 하지만 골다공증환자는 심각한 중증이거나, 골절이 된 뒤에야 의료보험 혜택을 받을 수 있습니다. 증상이 없는 초기에 적극적인 치료를 유도해야하는데, 우리의 의료현실은 그렇지 않습니다.● 원영준(관동대 의대 명지병원) : “치료 쪽에 집중되고 사고가 일어난 다음에 사후처리에 집중되고 이런 거 아닌가 생각됩니다. 이 질병 하나만 얘기하기보다는 우리가 가진 의식세계가 그 정도가 된 거 아닌가” 예방과 사전관리가 가장 절실한 계층은 역시 하위계층 독거노인들입니다. 80대 독거노인들의 건강실태를 파악하기 위해 방문간호사들이 찾아왔습니다.● 인터뷰 : (행복하다고 느끼시는

In [969]:
error

[['해운대 자생한방병원 김상돈 병원장'],
 ['의료사회복지사를 준비하는 대학생들을 위한 '],
 ['[웰빙라이프]발 빠른 우리 아이'],
 ['온누리스마일안과, '],
 [' ']]

In [989]:
article_raw = pd.DataFrame(zip(days, title, body))
article_raw.reset_index(inplace=True)
article_raw.columns = ['idx', 'day', 'title', 'body']
article_raw

idx        day                            title  \
0            0 2007-06-27          노후 10년은 병치레‥'건강수명' 관심부족   
1            1 2007-06-27      한국 출산율 1.19명 불과…세계 4번째 저출산국   
2            2 2007-06-27       2030년 亞.阿 도시 인구 2배 <UNFPA>   
3            3 2007-06-27                     영아 사망률 세계 최저   
4            4 2007-06-27  [쑥쑥 키 클리닉] 패스트푸드 즐기면 일찍 성장 멈출수도   
...        ...        ...                              ...   
530475  530475 2019-09-06      암 재발의 근원 암줄기세포만 콕 찍어주는 형광물질   
530476  530476 2004-09-25         콘플레이크·라면 이어 비스킷에서도 해충 발견   
530477  530477 2004-09-25        컵라면 과자 아이스크림서 '벌레' 나왔다 -1   
530478  530478 2004-09-25           ‘새집증후군’…교실 발암물질 기준치 8배   
530479  530479 2004-09-25                   추석연휴, 응급환자 이렇게   

                                                     body  
0       [뉴스데스크]● 앵커： 요즘 우리 평균수명은 많이 늘었지만 몸이 불편해 고생하시는 ...  
1       2007년 세계 인구‘는 지난해보다 7560만명이 늘어난 66억1590만명으로 집계...  
2       30년후 17억명 추가.. UNFPA "사고의 혁명 겪어야" (제네바=연합뉴스) 이...  
3       [한겨레] 우리나라의 영아 사망률이 세계에서 가장 낮은 수준인 것으로 나타났다. 2...  
4       50세 이상의 어른은 어릴 적 기억 속에 미국인은 높은 콧대에 거인, 심지어는 괴물...  
...                                                   ...  
530475  종양이 유도된 생쥐의 폐에서 타이니어가 종양근원세포를 붉은 색으로 물들임을 확인할 ...  
530476  [오마이뉴스 장윤선 기자] ▲ 벌레가 박혀있던 비스킷. 소비자는 제조사인 롯데제과측...  
530477  【서울=뉴시스】 최근 모 인스턴트 식품에서 벌레가 나왔다는 보도 이후 유통기한이 지...  
530478  ‘새집증후군’이 학교에까지 널리 퍼져 있어 대책 마련이 시급하다는 지적이다. 심각한...  
530479  <앵커> 추석연휴, 집안에 노약자나 어린이가 있는 경우는 반드시 1339 번을 기억...  

[530480 rows x 4 columns]

In [1013]:
article_raw_dropped = article_raw[['title','body']].drop_duplicates()

with open('article_raw_dropped.pickle', 'wb') as f: 
    pickle.dump(article_raw_dropped, f)
    
article_raw_dropped.to_csv('article_raw_dropped.tsv', sep='\t')

In [4]:
# pickle 불러오기
with open('./health_article/article_raw.pkl', 'rb') as f: 
    article_raw = pickle.load(f)

In [990]:
# pickle 저장
with open('article_raw.pickle', 'wb') as f: 
    pickle.dump(article_raw, f)

In [991]:
# tsv 저장
article_raw.to_csv('article_raw.tsv', sep='\t')

In [1014]:
body = article_raw_dropped.reset_index().body
body 

0         [뉴스데스크]● 앵커： 요즘 우리 평균수명은 많이 늘었지만 몸이 불편해 고생하시는 ...
1         2007년 세계 인구‘는 지난해보다 7560만명이 늘어난 66억1590만명으로 집계...
2         30년후 17억명 추가.. UNFPA "사고의 혁명 겪어야" (제네바=연합뉴스) 이...
3         [한겨레] 우리나라의 영아 사망률이 세계에서 가장 낮은 수준인 것으로 나타났다. 2...
4         50세 이상의 어른은 어릴 적 기억 속에 미국인은 높은 콧대에 거인, 심지어는 괴물...
                                ...                        
530475    종양이 유도된 생쥐의 폐에서 타이니어가 종양근원세포를 붉은 색으로 물들임을 확인할 ...
530476    [오마이뉴스 장윤선 기자] ▲ 벌레가 박혀있던 비스킷. 소비자는 제조사인 롯데제과측...
530477    【서울=뉴시스】 최근 모 인스턴트 식품에서 벌레가 나왔다는 보도 이후 유통기한이 지...
530478    ‘새집증후군’이 학교에까지 널리 퍼져 있어 대책 마련이 시급하다는 지적이다. 심각한...
530479    <앵커> 추석연휴, 집안에 노약자나 어린이가 있는 경우는 반드시 1339 번을 기억...
Name: body, Length: 528366, dtype: object

In [ ]:
제3조제3항제3호 2005.1.30.


의료 법률 : 특이패턴 반복되는거 의미있을거 불필요 
    지울건지 대체할건지 
    전처리
    문장단위 스플릿
    데이셋 형태로 가공

In [1178]:
# 특수문자 등 클린징
def cleansing(text):
    # 숫자의 경우 우선 생략          |(\d+) #숫자
    pattern = """ (\[.*\]) | (\<.*\>) | (【\w+=\w+】) #괄호문자
          |(\d+-\d+-\d+) #전화번호1
          |(\d+\)\d+-\d+) #전화번호2
          |(\w+\@.*\.\w+) #이메일주소
          |(www\.\w+\.[^\s+][\.\w]+) # 링크주소1
          |(http.*[a-zA-Z]) #링크주소2
          |(\w+=\w+) #글자=글자 형태
          |(\w*\s?기자) #기자이름
          |(\d+\.) #숫자목록
          |(\d+\w*\,?\d*원) #금액
          |(\s){2,}
          |[“”:，]
          |[■▦◆▲☎●▶▽△◇→━©ⓒ①②③④⑤⑥⑦®㈜※] #특수문자
          |[(\xa0)] #특수문자2
          |[!#$%&^\'()*+,<=>?@[\\]^_\"{\}~(\/)(\\\\)] #일반적 문장부호
          """

    p = re.compile(pattern, re.VERBOSE)
    tmp = p.sub("", text)
    tmp = tmp.strip()
    return tmp


# 영어 P. R. O. O. F. 와 같은 형식을 PROOF로 변경
def change_eng(text):
    tmp = text
    pattern_eng = "([a-zA-Z]\.\s*)"
    pp = re.compile(pattern_eng)

    find_iter = [f for f in pp.finditer(tmp)]
    tmp_copy = tmp

    for i in range(len(find_iter)):  
        origin = find_iter[i].group()
        changed = find_iter[i].group().replace('. ', '')
        tmp_copy = re.sub(origin,changed,tmp_copy)


    return tmp_copy

def preprocessing(text):
    tmp = text
    tmp = cleansing(tmp)
    tmp = change_eng(tmp)
    tmp = [ t.strip()+'.' for t in tmp.split('.') if t != '' ]
    
#     try:
#         tmp.index('..')
#         drop_id = tmp.index('..')
#         tmp.pop(drop_id)
#     except :
#         pass
    return tmp
    
    
all_article_list = []
all_article_seg = []

for i in tqdm(range(len(body))):
    cleansed_body = preprocessing(body[i])
    all_article_list.extend(cleansed_body)
    all_article_seg.extend([i]*len(cleansed_body))


print(len(all_article_list), len(all_article_seg))
    
# tmp = [ t.strip()+'.' for t in tmp.split('.')]
# |['!"#$%&\'()*+,-/:;<=>?@[\\]^_`{|}~']
# . - / ; : 이런것들은 마침표가 사라짐

100%|████████████████████████████████████████████████████████████████████████| 528366/528366 [03:33<00:00, 2472.37it/s]

8554103 8554103


In [1180]:
all_article = pd.DataFrame(zip(all_article_seg, all_article_list))
all_article.columns = ['segment', 'sentence']

In [1203]:
print(all_article.segment[-1:])
all_article.head()

8554102    528365
Name: segment, dtype: int64


segment                                           sentence
0        0  앵커： 요즘 우리 평균수명은 많이 늘었지만 몸이 불편해 고생하시는 노인들 주변에 많...
1        0          그냥 오래 사는 게 아니라 건강하게 오래 사는 이른바 건강수명이 중요하죠.
2        0                             우리의 현 실태를 가 집중 취재했습니다.
3        0  한국인의 평균수명은 현재  5세 OECD 국가의 평균치인 79세를 거의 따라 잡았습니다.
4        0  그러나 사고나 질병으로 거동이 불편한 기간을 뺀 건강히 사는 기간 즉 건강수명은 그...

In [14]:
with open('./health_article/all_article.pkl', 'wb') as f: 
    pickle.dump(all_article, f)

FileNotFoundError: [Errno 2] No such file or directory: './health_article/all_article.pkl'

In [8]:
with open('./health_article/all_article.pkl', 'rb') as f: 
    all_article = pickle.load(f)

In [13]:
input_data = pd.DataFrame()

sent_1 = []
sent_2 = []

article_num = int(528365)

for i in tqdm(range(article_num)):
    sentences = all_article[all_article.segment == i] #기사별로 문장을 구분

    sent_1.extend(list(sentences.sentence.values)) #현재문장
    sent_2.extend(list(sentences[1:].sentence.values)) #다음문장
    sent_2.append(sent_2[-1]) #마지막 문장 복사
    

# pd.DataFrame(zip(sent_1, sent_2))

100%|██████████████████████████████████████████████████████████████████████████| 528365/528365 [45:28<00:00, 193.68it/s]


In [14]:
health_news_inputdata = pd.DataFrame(zip(sent_1, sent_2))

In [23]:
with open('./health_news_inputdata.pkl', 'wb') as f :
    pickle.dump(health_news_inputdata, f)

In [ ]:
with open('./health_news_inputdata.pkl', 'rb') as f:
    health_news_inputdata = pickle.load(f)

In [21]:
health_news_inputdata.columns = ['Sentence_1', 'Sentence_2']
health_news_inputdata.reset_index(inplace=True)
health_news_inputdata.loc[:42]

# chatbot 같은 것은 인터뷰 형태의 기사들도 좋은 데이터가 될 것 같음 

index                                         Sentence_1  \
0       0  앵커： 요즘 우리 평균수명은 많이 늘었지만 몸이 불편해 고생하시는 노인들 주변에 많...   
1       1          그냥 오래 사는 게 아니라 건강하게 오래 사는 이른바 건강수명이 중요하죠.   
2       2                             우리의 현 실태를 가 집중 취재했습니다.   
3       3  한국인의 평균수명은 현재  5세 OECD 국가의 평균치인 79세를 거의 따라 잡았습니다.   
4       4  그러나 사고나 질병으로 거동이 불편한 기간을 뺀 건강히 사는 기간 즉 건강수명은 그...   
5       5  최근 조사에 따르면 건강수명은  6세 즉 말년에 10년 가량은 병상에 있다가 삶을 ...   
6       6  강은정 박사보건사회연구원  일반대중을 위해 장애를 예방한다든지 장애상태에서 일상생활...   
7       7  우리의 건강수명은 30개 OECD 국가가운데 24위로 우리보다 못한 나라는 체코 헝...   
8       8           덴마크 미국 등은 우리보다 평균수명은 짧지만 건강수명은 오히려 앞섭니다.   
9       9                       골다공증을 앓다가 허리 골절로 입원한 할머니입니다.   
10     10           골절이 심해 치료를 마쳐도 정상적인 상태로 돌아가기는 어려운 실정입니다.   
11     11                       이말순68세 골다공증 환자  여섯 대가 나갔다든가.   
12     12                                       그때는 무척 아팠어요.   
13     13  이처럼 노인들의 골다공증은 골절로 이어질 우려가 크기 때문에 초기 치료와 관리가 절...   
14     14  하지만 골다공증환자는 심각한 중증이거나 골절이 된 뒤에야 의료보험 혜택을 받을 수 ...   
15     15    증상이 없는 초기에 적극적인 치료를 유도해야하는데 우리의 의료현실은 그렇지 않습니다.   
16     16  원영준관동대 의대 명지병원  치료 쪽에 집중되고 사고가 일어난 다음에 사후처리에 집...   
17     17  이 질병 하나만 얘기하기보다는 우리가 가진 의식세계가 그 정도가 된 거 아닌가 예방...   
18     18           80대 독거노인들의 건강실태를 파악하기 위해 방문간호사들이 찾아왔습니다.   
19     19                            인터뷰  행복하다고 느끼시는 편이에요 예.   
20     20                               하하하 무슨 말인지도 모르고 대답해.   
21     21                                       행복하긴 뭐가 행복해.   
22     22         여기저기 아픈 곳이 많은 할머니들 그런데도 건강검진을 받아본 적은 없습니다.   
23     23                                      돈이 없어서가 아닙니다.   
24     24  기초생활수급대상자인 이 노인들은 매년 무료로 건강검진을 받을 수 있건만 혼자 살고 ...   
25     25                         난생 처음 보건소에서 건강진단을 받게 됐습니다.   
26     26                  인터뷰  저녁도 안먹고 저녁은 잡수시고 아침만 굶고 오세요.   
27     27                               언니랑 가면 되겠네 난 혼자는 못가.   
28     28           정부는 올 3월부터 방문간호를 통해 독거노인들의 안부를 살피고 있습니다.   
29     29  집에서 거리에서 노인들을 만나 상태를 파악하고 있지만 그 손길을 기다리는 쪽에서 보...   
30     30  이승욱 원장 서울대 보건대학원  말 못하고 행동 못하고 그러면 그때부터 건강하지 않...   
31     31                                  건강수명을 늘리기 위한 행동은.   
32     32  평균수명을 끌어올리기 위해 수십 년 허겁지겁 뛰어온 대한민국 이제 건강수명을 늘리기...   
33     33                앵커：  요즘 그렇게들 건강에 신경 쓴다지만 실상은 의외로군요.   
34     34                                             그렇습니다.   
35     35                       건강에 대해 관심이 높다고 하지만 아직 부족합니다.   
36     36               우리의 생활습관을 재미교포와 비교한 흥미로운 연구결과가 있습니다.   
37     37              음주 흡연 비만도 모두 재미교포에 비해 미흡한 것으로 확인됐습니다.   
38     38  한국 사람이 한국에서 살 때보다 미국에서 생활할 때 건강에 더 신경을 쓴다는 얘기죠...   
39     39  3년 안에 서너 살 올려야 한다는 얘기인데요 2000년 66세에서 2005년에  6...   
40     40  개인의 행복도 행복이지만 요즘 같은 저출산 고령화 사회에서 은퇴시점을 늦추고 노동인...   
41     41                                      앵커：  잘 들었습니다.   
42     42  2007년 세계 인구‘는 지난해보다 7560만명이 늘어난 66억1590만명으로 집계됐다.   

                                           Sentence_2  
0           그냥 오래 사는 게 아니라 건강하게 오래 사는 이른바 건강수명이 중요하죠.  
1                              우리의 현 실태를 가 집중 취재했습니다.  
2   한국인의 평균수명은 현재  5세 OECD 국가의 평균치인 79세를 거의 따라 잡았습니다.  
3   그러나 사고나 질병으로 거동이 불편한 기간을 뺀 건강히 사는 기간 즉 건강수명은 그...  
4   최근 조사에 따르면 건강수명은  6세 즉 말년에 10년 가량은 병상에 있다가 삶을 ...  
5   강은정 박사보건사회연구원  일반대중을 위해 장애를 예방한다든지 장애상태에서 일상생활...  
6   우리의 건강수명은 30개 OECD 국가가운데 24위로 우리보다 못한 나라는 체코 헝...  
7            덴마크 미국 등은 우리보다 평균수명은 짧지만 건강수명은 오히려 앞섭니다.  
8                        골다공증을 앓다가 허리 골절로 입원한 할머니입니다.  
9            골절이 심해 치료를 마쳐도 정상적인 상태로 돌아가기는 어려운 실정입니다.  
10                       이말순68세 골다공증 환자  여섯 대가 나갔다든가.  
11                                       그때는 무척 아팠어요.  
12  이처럼 노인들의 골다공증은 골절로 이어질 우려가 크기 때문에 초기 치료와 관리가 절...  
13  하지만 골다공증환자는 심각한 중증이거나 골절이 된 뒤에야 의료보험 혜택을 받을 수 ...  
14    증상이 없는 초기에 적극적인 치료를 유도해야하는데 우리의 의료현실은 그렇지 않습니다.  
15  원영준관동대 의대 명지병원  치료 쪽에 집중되고 사고가 일어난 다음에 사후처리에 집...  
16  이 질병 하나만 얘기하기보다는 우리가 가진 의식세계가 그 정도가 된 거 아닌가 예방...  
17           80대 독거노인들의 건강실태를 파악하기 위해 방문간호사들이 찾아왔습니다.  
18                            인터뷰  행복하다고 느끼시는 편이에요 예.  
19                               하하하 무슨 말인지도 모르고 대답해.  
20         

#현재 issue 들

 - 띄어쓰기 2번씩 있는 애들 가끔 있는데 걸러지지 않음

replace 말고 regular expression 사용
문장 시작에 특수기호 '(문자, 특문, 영어)' 시작 부분 삭제
예)[마이데일리 =. 00칼럼], [관련기사] 

마지막 문장에 출처 등 제거
마지막 문단 어떻게 잘라낼 거냐? 맨 끝문장 + 영어 및 특문 포함인 경우
〈이재동/경희대 한의학과 교수〉- 대한민국 희망언론! 경향신문, 구독신청(http://smile. khan. co. kr) - ⓒ 경향신문 & 미디어칸(www. khan. co. kr), 무단전재 및 재배포 금지〈미디어칸은 한국온라인신문협회(www. kona. or. kr)의 디지털뉴스이용규칙에 따른 저작권을 행사합니다.〉'

이후 전처리 

이메일 형식: aaa.aaa.aa.aa, http://, https:// --> 주소
전화번호: 00)000-0000, 000-0000-0000  --> 전화번호
특수기호: ⓒ&

숫자, 숫자.숫자 --> [숫자]로 대체 

패턴에 따라 불필요한 것 대체 필요
sentence로 나눈 다음에
기본적인 puctuation: '!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'(string.punctuation)

특문
줄임말이나 또는 표시 등의 문장부호는 문맥을 파악하는데 중요하지 않나? 꼭 없애야 하는가 : ·…
숫자는 잘못된 편향 학습 때문에 지우는게 맞는것 같음
단위는 없애지 않음 --> 이것도 기호화?
()는 동의어 또는 영문명, 준말 등을 표현하기도 하는 기호 없애나?

숫자 -> [number]
"\d+" or "\d+.\d+"
re.sub(pattern, "[number]", body[530479])
숫자, (), [], <>, ■ 마지막에 위치(sentence로 끊었을때 마지막 위치) 숫자.숫자 \xa0 ▲ 00)000-0000 000-0000-0000 -

In [ ]:
"""
6,726개 날짜별 기사 파일(파일 내 다수 기사 존재)
중복을 제거한 기사 총 528,366개 (원래 530,480개)
sentence 단위로 분리한 문장 총 8,554,103

"""